From 09bde402e9284b59a8ea39c7df3ecf3d71479d51 Mon Sep 17 00:00:00 2001
From: Christopher Albert <albert@tugraz.at>
Date: Tue, 31 Mar 2026 08:26:57 +0200
Subject: [PATCH] fortran: Fix ICE in gfc_trans_create_temp_array for
 assumed-rank [PR100194]

When a non-contiguous assumed-rank actual argument is passed to a
contiguous assumed-rank dummy, the compiler routes it through
gfc_conv_subref_array_arg which uses the scalarizer.  The scalarizer
requires known rank at compile time, but assumed-rank arrays have
rank = -1, hitting gcc_assert (ss->dimen > 0).

Skip the scalarizer path for assumed-rank expressions and let them
fall through to gfc_conv_array_parameter, which handles assumed-rank
via the runtime pack/unpack functions.

gcc/fortran/ChangeLog:

	PR fortran/100194
	* trans-expr.cc (gfc_conv_procedure_call): Skip
	gfc_conv_subref_array_arg for assumed-rank actual arguments
	(e->rank == -1) when the dummy is contiguous.

gcc/testsuite/ChangeLog:

	PR fortran/100194
	* gfortran.dg/pr100194.f90: New test.

Signed-off-by: Christopher Albert <albert@tugraz.at>
---
 gcc/fortran/trans-expr.cc              |  3 ++-
 gcc/testsuite/gfortran.dg/pr100194.f90 | 16 ++++++++++++++++
 2 files changed, 18 insertions(+), 1 deletion(-)
 create mode 100644 gcc/testsuite/gfortran.dg/pr100194.f90

diff --git a/gcc/fortran/trans-expr.cc b/gcc/fortran/trans-expr.cc
index d5254be007d..52918961584 100644
--- a/gcc/fortran/trans-expr.cc
+++ b/gcc/fortran/trans-expr.cc
@@ -7956,7 +7956,8 @@ gfc_conv_procedure_call (gfc_se * se, gfc_symbol * sym,
 		       && (fsym->attr.target
 			   ? gfc_is_not_contiguous (e)
 			   : !gfc_is_simply_contiguous (e, false, true))
-		       && gfc_expr_is_variable (e))
+		       && gfc_expr_is_variable (e)
+		       && e->rank != -1)
 		{
 		  gfc_conv_subref_array_arg (&parmse, e, nodesc_arg,
 					     fsym->attr.intent,
diff --git a/gcc/testsuite/gfortran.dg/pr100194.f90 b/gcc/testsuite/gfortran.dg/pr100194.f90
new file mode 100644
index 00000000000..09eb3db94a3
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/pr100194.f90
@@ -0,0 +1,16 @@
+! { dg-do compile }
+!
+! PR fortran/100194
+! ICE in gfc_trans_create_temp_array when passing a non-contiguous
+! assumed-rank array to a contiguous assumed-rank dummy argument.
+!
+! Contributed by Martin Diehl <m.diehl@kit.edu>
+
+subroutine s(x)
+   real :: x(..)
+   call t(x)
+contains
+   subroutine t(y)
+      real, contiguous :: y(..)
+   end
+end
-- 
2.53.0

